lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,454 @@
+              auto memPort = sub.getInput();
               auto type = memPort.getType().cast<BundleType>();
               auto enableFieldId =
                   type.getFieldID((unsigned)ReadPortSubfield::en);
               auto dataFieldId =
                   type.getFieldID((unsigned)ReadPortSubfield::data);
               auto addressFieldId =
                   type.getFieldID((unsigned)ReadPortSubfield::addr);
-              llvm::SmallDenseSet<size_t> fieldIdsToAdd = {
-                  enableFieldId, dataFieldId, addressFieldId};
-              gatherAggregateLeafs(memPort, rootValues, fieldIdsToAdd);
-
-              size_t baseId;
-              if (!getStorageLocIdFromVal(memPort, baseId))
-                continue;
-              Value dataField;
-              if (!getValueFromStorageLocId(baseId + dataFieldId, dataField))
+              memPorts.insert(memPort);
+              if (fieldIndex == enableFieldId || fieldIndex == dataFieldId ||
+                  fieldIndex == addressFieldId) {
+                setValRefsTo(memPort, Node(memPort, 0));
+              } else
+                return;
+            }
+            forallRefersTo(
+                sub.getInput(),
+                [&](Node subBase) {
+                  isValid = true;
+                  setValRefsTo(res, Node(subBase.getValue(),
+                                         subBase.getFieldID() + fieldIndex));
+                  return success();
+                },
+                false);
+            if (isValid)
+              worklist.push_back(res);
+          })
+          .Case<SubindexOp>([&](SubindexOp sub) {
+            auto res = sub.getResult();
+            bool isValid = false;
+            auto index = sub.getIndex();
+            forallRefersTo(
+                sub.getInput(),
+                [&](Node subBase) {
+                  isValid = true;
+                  setValRefsTo(res, Node(subBase.getValue(),
+                                         subBase.getFieldID() + index + 1));
+                  return success();
+                },
+                false);
+            if (isValid)
+              worklist.push_back(res);
+          })
+          .Case<SubaccessOp>([&](SubaccessOp sub) {
+            auto vecType = sub.getInput().getType().cast<FVectorType>();
+            auto res = sub.getResult();
+            bool isValid = false;
+            forallRefersTo(
+                sub.getInput(),
+                [&](Node subBase) {
+                  isValid = true;
+                  // The result of a subaccess can refer to multiple storage
+                  // locations corresponding to all the possible indices.
+                  for (size_t index = 0; index < vecType.getNumElements();
+                       ++index)
+                    setValRefsTo(
+                        res,
+                        Node(subBase.getValue(),
+                             subBase.getFieldID() + 1 +
+                                 index *
```

Comment:
More clang-format fun, although this makes more sense.  This looks like output from `pretty::PrettyPrinter` :joy:  :wink: .